src/main/java/com/princexml/wrapper/AbstractPrince.java

```diff
@@ -173,8 +173,8 @@ protected boolean readMessages(BufferedReader reader) throws IOException {
 
                 if (events != null && msgTag.equals("msg")) {
                     handleMessage(msgBody);
-                } else if (msgTag.equals("dat")) {
-                    // TODO Data messages?
+                } else if (events != null && msgTag.equals("dat")) {
+                    handleDataMessage(msgBody);
```

Comment:
Perhaps slightly clearer to test events on a nested if, just so the outer if is basically switching on msgTag.

Reply:
I've nested the `events` check and used a switch for the `msgTag` check.